avr-hal-generic/src/lib.rs

```diff
@@ -1,6 +1,38 @@
 #![no_std]
 #![cfg_attr(avr_hal_asm_macro, feature(asm_experimental_arch))]
 #![cfg_attr(not(avr_hal_asm_macro), feature(llvm_asm))]
+//! Defines the internal API used to generate the per-device external API
+//! (user-facing) in the HAL crates.
+//!
+//! # Macro Design (WIP)
+//!
+//! As part of an effort to make the macros here more maintainable and less
+//! repetitive, they are being restructured module-by-module. Currently revised
+//! modules are:
+//! - [`adc`]
+//! - [`port`]
+//!
+//! The general guiding principals are:
+//! 1. What goes inside the macro invocation should look like regular code as
+//! much as possible;
```

Comment:
To give an example from your code here where this topic hits: The syntax suggests the existence of an `AdcProvider` trait - but it is not defined anywhere.  This can cost a lot of time to figure out if someone isn't immediately aware of it just being "nice syntax".

Reply:
Is this still a problem if the items inside the macro actually match onesthat end up existing? For instance, if AdcProvider is replaced by AdcChannel. The guideline would then be "the input to the macro must resemble/indicate the items that come out, but mostly those of the user-facing API". Because it seems like a common pattern across the codebase that each module defines a low level trait which is used along with higher level wrappers, and the macros define the low level implementation from the data passed to them. So the input should resemble the final layout but allow setting the underlying register operations. The generated `*Ops` impls can be implicit.